torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -766,14 +766,13 @@ std::shared_ptr<at::cuda::CUDAEvent> ProcessGroupNCCL::CUDAEventCache::create(
     std::lock_guard<std::mutex> lock(cacheMutex_);
     auto& events = eventsArray_[timing ? 1 : 0];
     if (!events.empty()) {
-      event = events.back();
-      events.pop_back();
+      event = events.front();
+      events.pop_front();
```

Comment:
there isn't anything wrong with using a deque, but I think its just as good as using a vector.

If we think its "safer" to reuse an older event before reusing a newer event, that suggests we do not have a good cache design.  We must be 100% sure that as soon as an event is put into the cache, it is safe to immediately reuse.  If we are sure of this, then changing the reuse order is pointless.  If not, we can't safely use a cache even with LIFO order.

Reply:
That is true. So if we think the deleter will be called at least after we call erase in watchdog, then yes, we are 100% sure it's safe even for stack. Originally I was not sure if "making it safer" is a valid argument. So I guess just conceptually, FIFO is better?

Reply:
I think conceptually they are equally good, so probably just leave it alone. If you want to land the deque change I am fine with it but you should update your PR desc to just say you like deque better instead of suggesting it is safer :)

Reply:
oh, got it. Sure, let me update the description.